File: server/core/events.py
from pydantic import BaseModel
from typing import Dict, Any, Literal
from datetime import datetime

class GraphEvent(BaseModel):
    type: Literal["NodeAdded", "EdgeAdded", "NodeUpdated", "EdgeRemoved"]
    timestamp: datetime = datetime.now()

class NodeAdded(GraphEvent):
    type: Literal["NodeAdded"] = "NodeAdded"
    node_id: str
    label: str
    properties: Dict[str, Any]

class EdgeAdded(GraphEvent):
    type: Literal["EdgeAdded"] = "EdgeAdded"
    edge_id: int
    source_id: str
    target_id: str
    label: str
    properties: Dict[str, Any]

class NodeUpdated(GraphEvent):
    type: Literal["NodeUpdated"] = "NodeUpdated"
    node_id: str
    label: str
    properties: Dict[str, Any]

class EdgeRemoved(GraphEvent):
    type: Literal["EdgeRemoved"] = "EdgeRemoved"
    edge_id: int
    source_id: str
    target_id: str

# Event bus (in-memory for now, can be replaced with Redis later)
active_connections = []

async def broadcast_event(event: GraphEvent):
    """Broadcast event to all connected WebSocket clients"""
    for connection in active_connections[:]:
        try:
            await connection.send_json(event.model_dump())
        except:
            active_connections.remove(connection)

File: client/src/types/events.ts
export type GraphEvent = 
  | NodeAdded 
  | EdgeAdded 
  | NodeUpdated 
  | EdgeRemoved;

export interface NodeAdded {
  type: "NodeAdded";
  node_id: string;
  label: string;
  properties: Record<string, any>;
  timestamp: string;
}

export interface EdgeAdded {
  type: "EdgeAdded";
  edge_id: number;
  source_id: string;
  target_id: string;
  label: string;
  properties: Record<string, any>;
  timestamp: string;
}

export interface NodeUpdated {
  type: "NodeUpdated";
  node_id: string;
  label: string;
  properties: Record<string, any>;
  timestamp: string;
}

export interface EdgeRemoved {
  type: "EdgeRemoved";
  edge_id: number;
  source_id: string;
  target_id: string;
  timestamp: string;
}
